### PR TITLE
Add items for project map layers and relations to expression builder 

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows easy insertion of map layer IDs and relation IDs into expressions. Numerous expression functions now utilise these, so it makes sense to allow them to be easily inserted.

Fix #11680, #16879

Sponsored by Andreas Neumann

### CODE REVIEW

File: src/gui/qgsexpressionbuilderwidget.cpp [hunk 1]
```diff
@@ -440,6 +443,26 @@ void QgsExpressionBuilderWidget::loadRecent( const QString &collection )
   }
 }
 
+void QgsExpressionBuilderWidget::loadLayers()
+{
+  QMap<QString, QgsMapLayer *> layers = QgsProject::instance()->mapLayers();
```
Comment (m-kuhn) [inline src/gui/qgsexpressionbuilderwidget.cpp]: If you replace the `QgsProject::instance()` call with a setter you help to make the future brighter.

Comment (nyalldawson) [inline src/gui/qgsexpressionbuilderwidget.cpp]: Ok done, but I've left it to default to QgsProject::instance(). This widget (and the accompanying dialog) is scattered all through gui/app. It's not a trivial task to catch all those uses and flip them to manually setting the project)

Comment (m-kuhn) [inline src/gui/qgsexpressionbuilderwidget.cpp]: Yes, that's what's scaring me as well when I think that we ever would want to get rid of `QgsProject::instance`.

Comment (nyalldawson) [inline src/gui/qgsexpressionbuilderwidget.cpp]: Well.. realistically it's unlikely that we'll even remove it from core for 3.0. So removing it from GUI is a 4.0+ goal. 

But anyway, we should stay on top of this (like you've done here) and try to make it easier for our future selves...

**Pull request merged**